example-app/App.tsx

```diff
@@ -40,6 +40,55 @@ const Header = () => {
 };
 
 const App = () => {
+  const [result, setResult] = React.useState<
+    | {
+        success: true;
+      }
+    | {
+        success: false;
+        errorReason: string;
+      }
+    | undefined
+  >(undefined);
+
+  React.useEffect(() => {
+    Linking.addEventListener('url', event => {
+      setResult(handleDeepLink(event.url));
+    });
+
+    async function getDeepLink() {
+      let url = await Linking.getInitialURL();
+      let res = handleDeepLink(url);
+      setResult(res);
+    }
+
+    getDeepLink();
+  }, []);
+
+  const handleDeepLink = (
+    url: string | null,
+  ):
+    | {
+        success: true;
+      }
+    | {
+        success: false;
+        errorReason: string;
+      }
+    | undefined => {
```

Comment:
It might be an idea to define a type alias instead of having the code everywhere

```typescript
type DeepLinkResult =
  | {
    success: true;
    }
  | {
...
```